File: src/jetnet/session.ts
/**
 * session.ts -- JETNET session helper (Node / TypeScript)
 *
 * Drop-in equivalent of session.py for Node / Next.js / Bun projects.
 *
 * Usage:
 *   import { login, ensureSession, jetnetRequest } from './session'
 *
 *   const session = await login()                        // reads env vars
 *   const data    = await jetnetRequest('GET',
 *                     '/api/Aircraft/getRegNumber/N12345/{apiToken}',
 *                     session)
 *
 * Environment variables (same names as Python helper):
 *   JETNET_EMAIL     -- emailAddress (capital A matters on the server)
 *   JETNET_PASSWORD  -- password
 *   JETNET_BASE_URL  -- default: https://customer.jetnetconnect.com
 */

const DEFAULT_BASE_URL = 'https://customer.jetnetconnect.com'

export interface SessionState {
  baseUrl: string
  email: string
  password: string
  bearerToken: string
  apiToken: string
  createdAt: number
  lastValidatedAt: number
}

export interface JetnetErrorBody {
  endpoint: string
  status: string
  detail?: string
}

export class JetnetError extends Error {
  endpoint: string
  status: string

  constructor(endpoint: string, status: string, detail?: string) {
    super(`JETNET error [${endpoint}]: ${status}${detail ? ' -- ' + detail : ''}`)
    this.name = 'JetnetError'
    this.endpoint = endpoint
    this.status = status
  }
}

function nowSeconds(): number {
  return Date.now() / 1000
}

/**
 * Detect JETNET application-level errors in a parsed response JSON.
 * JETNET uses two error styles:
 *   1. responsestatus starting with "ERROR" or "INVALID" (most endpoints)
 *   2. RFC 7807: { "title": "...", "status": 400, "detail": "..." }
 */
function normalizeError(
  responseJson: Record<string, unknown>,
  endpoint = ''
): JetnetError | null {
  const rs = responseJson['responsestatus']
  if (typeof rs === 'string') {
    const upper = rs.toUpperCase()
    if (upper.startsWith('ERROR') || upper.startsWith('INVALID')) {
      return new JetnetError(endpoint, rs)
    }
  }

  if ('title' in responseJson && 'status' in responseJson) {
    const title = String(responseJson['title'] ?? '')
    const detail = responseJson['detail'] ? String(responseJson['detail']) : undefined
    return new JetnetError(endpoint, title, detail)
  }

  return null
}

const TOKEN_TTL_SECONDS = 50 * 60  // 50 min (tokens last ~60 min)

/**
 * Authenticate and return a new SessionState.
 *
 * CRITICAL: The field name is `emailAddress` with a capital A.
 * Sending `email` or `emailaddress` returns HTTP 200 with an error body.
 */
export async function login(
  email?: string,
  password?: string,
  baseUrl?: string
): Promise<SessionState> {
  const resolvedBaseUrl = baseUrl ?? process.env['JETNET_BASE_URL'] ?? DEFAULT_BASE_URL
  const resolvedEmail = email ?? process.env['JETNET_EMAIL']
  const resolvedPassword = password ?? process.env['JETNET_PASSWORD']

  if (!resolvedEmail || !resolvedPassword) {
    throw new Error(
      'JETNET credentials missing. Set JETNET_EMAIL and JETNET_PASSWORD env vars ' +
      'or pass email/password to login().'
    )
  }

  const url = `${resolvedBaseUrl}/api/Admin/APILogin`
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ emailAddress: resolvedEmail, password: resolvedPassword }),
  })

  if (!response.ok) {
    throw new Error(`JETNET login HTTP ${response.status}: ${await response.text()}`)
  }

  const data = (await response.json()) as Record<string, unknown>
  const err = normalizeError(data, 'APILogin')
  if (err) throw err

  const bearerToken = String(data['bearerToken'] ?? '')
  const apiToken = String(data['apiToken'] ?? '')

  if (!bearerToken || !apiToken) {
    throw new Error(
      'JETNET login returned no tokens. Check credentials. ' +
      `Response: ${JSON.stringify(data).slice(0, 200)}`
    )
  }

  const now = nowSeconds()
  return {
    baseUrl: resolvedBaseUrl,
    email: resolvedEmail,
    password: resolvedPassword,
    bearerToken,
    apiToken,
    createdAt: now,
    lastValidatedAt: now,
  }
}

/**
 * Call /getAccountInfo as a lightweight token health-check.
 * Returns the raw response dict.
 */
export async function getAccountInfo(
  session: SessionState
): Promise<Record<string, unknown>> {
  const url = `${session.baseUrl}/api/Admin/getAccountInfo/${session.apiToken}`
  const response = await fetch(url, {
    method: 'GET',
    headers: { Authorization: `Bearer ${session.bearerToken}` },
  })

  if (!response.ok) {
    throw new Error(`getAccountInfo HTTP ${response.status}`)
  }

  const data = (await response.json()) as Record<string, unknown>
  const err = normalizeError(data, 'getAccountInfo')
  if (err) throw err

  return data
}

/**
 * Return a valid session, refreshing if the token is stale.
 *
 * Strategy (three-tier defense):
 *   1. Age check: if token is < TOKEN_TTL old, return as-is.
 *   2. Probe /getAccountInfo; if it succeeds, update lastValidatedAt and return.
 *   3. Full re-login. If that fails, throw -- do not swallow the error.
 */
export async function ensureSession(
  session: SessionState
): Promise<SessionState> {
  const age = nowSeconds() - session.createdAt

  if (age < TOKEN_TTL_SECONDS) {
    return session
  }

  try {
    await getAccountInfo(session)
    return { ...session, lastValidatedAt: nowSeconds() }
  } catch {
    // Fall through to re-login
  }

  return await login(session.email, session.password, session.baseUrl)
}

/**
 * Make an authenticated JETNET request.
 *
 * - Substitutes `{apiToken}` in the path automatically.
 * - Retries once after a fresh login if the token appears invalid.
 * - Throws JetnetError for application-level errors.
 *
 * @param method   'GET' | 'POST'
 * @param path     Endpoint path, e.g. '/api/Aircraft/getRegNumber/N12345/{apiToken}'
 * @param session  Current SessionState
 * @param body     Optional POST body (will be JSON-serialized)
 */
export async function jetnetRequest(
  method: 'GET' | 'POST',
  path: string,
  session: SessionState,
  body?: Record<string, unknown>
): Promise<Record<string, unknown>> {
  const resolvedPath = path.replace('{apiToken}', session.apiToken)
  const url = `${session.baseUrl}${resolvedPath}`

  const headers: Record<string, string> = {
    Authorization: `Bearer ${session.bearerToken}`,
    'Content-Type': 'application/json',
  }

  const init: RequestInit = {
    method,
    headers,
    ...(body !== undefined ? { body: JSON.stringify(body) } : {}),
  }

  let response = await fetch(url, init)

  if (!response.ok) {
    throw new Error(`HTTP ${response.status} ${response.statusText} -- ${url}`)
  }

  let data = (await response.json()) as Record<string, unknown>
  let err = normalizeError(data, path)

  if (err && typeof err.status === 'string' && err.status.toUpperCase().startsWith('INVALID')) {
    const fresh = await login(session.email, session.password, session.baseUrl)
    Object.assign(session, fresh)

    const retryPath = path.replace('{apiToken}', fresh.apiToken)
    const retryUrl = `${fresh.baseUrl}${retryPath}`
    response = await fetch(retryUrl, {
      ...init,
      headers: { ...headers, Authorization: `Bearer ${fresh.bearerToken}` },
    })
    data = (await response.json()) as Record<string, unknown>
    err = normalizeError(data, path)
  }

  if (err) throw err
  return data
}

/**
 * Convenience: call ensureSession then jetnetRequest in one step.
 * Returns [updatedSession, responseData].
 */
export async function refreshAndRequest(
  method: 'GET' | 'POST',
  path: string,
  session: SessionState,
  body?: Record<string, unknown>
): Promise<[SessionState, Record<string, unknown>]> {
  const fresh = await ensureSession(session)
  const data = await jetnetRequest(method, path, fresh, body)
  return [fresh, data]
}

File: src/jetnet/session.py
"""
session.py -- JETNET API session manager

Handles login, token storage, health-checking via /getAccountInfo,
and auto-refresh on token expiry. Use this as the single source of
truth for all JETNET auth in your application.

Usage:
    from src.jetnet.session import login, ensure_session, jetnet_request

    session = login("you@example.com", "yourpassword")
    result  = jetnet_request("GET", "/api/Aircraft/getRegNumber/N12345/{apiToken}", session)
"""

from __future__ import annotations
import os
import time
import requests
from dataclasses import dataclass, field
from typing import Optional

BASE_URL = os.getenv("JETNET_BASE_URL", "https://customer.jetnetconnect.com")
TOKEN_TTL_SECONDS = int(os.getenv("JETNET_TOKEN_TTL", "3000"))  # 50 min (tokens last ~60 min)


@dataclass
class SessionState:
    base_url: str
    email: str
    password: str
    bearer_token: str
    api_token: str
    created_at: float = field(default_factory=time.time)
    last_validated_at: float = field(default_factory=time.time)

    def age_seconds(self) -> float:
        return time.time() - self.created_at

    def seconds_since_validated(self) -> float:
        return time.time() - self.last_validated_at

    def is_stale(self, ttl: int = TOKEN_TTL_SECONDS) -> bool:
        """True if the token has likely expired based on wall-clock age."""
        return self.age_seconds() > ttl


class JetnetError(Exception):
    """Raised when JETNET returns an application-level error (HTTP 200 + bad status)."""
    def __init__(self, message: str, endpoint: str = "", raw_status: str = ""):
        self.endpoint = endpoint
        self.raw_status = raw_status
        super().__init__(message)


def normalize_error(response_json: dict, endpoint: str = "") -> Optional[JetnetError]:
    """
    Detect application-level errors that arrive with HTTP 200.

    JETNET can signal errors in two ways:
      1. responsestatus starts with "ERROR" or contains "INVALID"
      2. RFC-style error body with a "title" or "type" field

    Returns a JetnetError if an error is detected, None otherwise.
    """
    status = response_json.get("responsestatus", "")
    if status:
        upper = status.upper()
        if upper.startswith("ERROR") or "INVALID" in upper:
            return JetnetError(
                f"JETNET error: {status}",
                endpoint=endpoint,
                raw_status=status,
            )

    if "title" in response_json or "type" in response_json:
        title = response_json.get("title", response_json.get("type", "Unknown error"))
        return JetnetError(
            f"API error: {title}",
            endpoint=endpoint,
            raw_status=str(response_json.get("status", "")),
        )

    return None


def login(
    email: Optional[str] = None,
    password: Optional[str] = None,
    base_url: Optional[str] = None,
) -> SessionState:
    """
    Authenticate with the JETNET API and return a SessionState.

    Reads JETNET_EMAIL / JETNET_PASSWORD / JETNET_BASE_URL from environment
    if not passed directly.

    IMPORTANT: emailAddress has a capital A -- this function handles that for you.
    """
    email    = email    or os.environ["JETNET_EMAIL"]
    password = password or os.environ["JETNET_PASSWORD"]
    url      = (base_url or BASE_URL).rstrip("/")

    r = requests.post(
        f"{url}/api/Admin/APILogin",
        json={"emailAddress": email, "password": password},
        timeout=30,
    )
    r.raise_for_status()
    data = r.json()

    err = normalize_error(data, endpoint="APILogin")
    if err:
        raise err

    return SessionState(
        base_url=url,
        email=email,
        password=password,
        bearer_token=data["bearerToken"],
        api_token=data.get("apiToken") or data.get("securityToken"),
    )


def get_account_info(session: SessionState) -> dict:
    """
    Call /getAccountInfo -- a lightweight, read-only probe endpoint.

    Use this to:
      - Verify a token is still valid before starting a long workflow
      - Observe practical token TTL (see scripts/token_probe.py)
      - Confirm the session is healthy after re-login

    Returns the account info dict, or raises JetnetError if the token is invalid.
    """
    r = requests.get(
        f"{session.base_url}/api/Admin/getAccountInfo/{session.api_token}",
        headers={"Authorization": f"Bearer {session.bearer_token}"},
        timeout=15,
    )
    r.raise_for_status()
    data = r.json()

    err = normalize_error(data, endpoint="getAccountInfo")
    if err:
        raise err

    session.last_validated_at = time.time()
    return data


def ensure_session(session: SessionState) -> SessionState:
    """
    Validate the session before starting any workflow. Auto-refreshes once on expiry.

    Call this at the start of any multi-step workflow (Golden Path, bulk export, etc.)
    to prevent mid-workflow token failures.

    Strategy:
      1. If token looks stale by age, skip the probe and re-login immediately.
      2. Otherwise call /getAccountInfo.
      3. If /getAccountInfo fails, re-login once and retry.
      4. If it still fails, raise -- credentials or service issue.
    """
    if session.is_stale():
        return login(session.email, session.password, session.base_url)

    try:
        get_account_info(session)
        return session
    except (JetnetError, requests.HTTPError):
        pass

    try:
        refreshed = login(session.email, session.password, session.base_url)
        get_account_info(refreshed)
        return refreshed
    except Exception as e:
        raise JetnetError(
            f"Session refresh failed. Check credentials and connectivity. Original error: {e}",
            endpoint="ensure_session",
        ) from e


def jetnet_request(
    method: str,
    path: str,
    session: SessionState,
    json: Optional[dict] = None,
    timeout: int = 60,
    auto_refresh: bool = True,
) -> dict:
    """
    Make an authenticated JETNET API request.

    Handles:
      - Bearer token in Authorization header
      - apiToken substitution in URL path
      - Application-level error detection (HTTP 200 but responsestatus = ERROR)
      - One automatic re-auth on INVALID SECURITY TOKEN

    Args:
        method:       "GET" or "POST"
        path:         e.g. "/api/Aircraft/getRegNumber/N12345/{apiToken}"
                      Use literal {apiToken} -- it will be substituted automatically.
        session:      SessionState from login() or ensure_session()
        json:         POST body dict (None for GET requests)
        timeout:      Request timeout in seconds
        auto_refresh: If True, re-login once on token error and retry

    Returns:
        Parsed JSON response dict

    Raises:
        JetnetError: on application-level errors
        requests.HTTPError: on HTTP 4xx/5xx errors
    """
    url = f"{session.base_url}{path}".replace("{apiToken}", session.api_token)
    headers = {
        "Authorization": f"Bearer {session.bearer_token}",
        "Content-Type": "application/json",
    }

    r = requests.request(method, url, headers=headers, json=json, timeout=timeout)
    r.raise_for_status()
    data = r.json()

    err = normalize_error(data, endpoint=path)
    if err:
        if auto_refresh and "INVALID" in err.raw_status.upper():
            refreshed = login(session.email, session.password, session.base_url)
            return jetnet_request(method, path, refreshed, json=json,
                                  timeout=timeout, auto_refresh=False)
        raise err

    return data


def refresh_session(session: SessionState) -> SessionState:
    """Force a fresh login regardless of token age. Returns a new SessionState."""
    return login(session.email, session.password, session.base_url)
